File: IP/3rd/sampleprogram/11-3-7.cpp
#define _CRT_SECURE_NO_WARNINGS
#define _USE_MATH_DEFINES
#include <iostream>
#include <cmath>
#include <opencv2/opencv.hpp>
using namespace std;
using namespace cv;
string win_src = "src";
string win_dst = "dst";

int main()
{
  string file_src = "src_depth_wall.png"; // 入力画像のファイル名
  string file_dst = "dst.png"; // 出力画像のファイル名
  // Mat img_src = imread(file_src, 1); // 入力画像（カラー）の読み込み
  Mat img_src = imread(file_src, 0); // 入力画像（グレースケール）の読み込み

  Mat img_dst;
  if(!img_src.data) {
    cout << "error" << endl;
    return -1;
  }

  // ウインドウ生成
  namedWindow(win_src, WINDOW_AUTOSIZE);
  namedWindow(win_dst, WINDOW_AUTOSIZE);

  img_dst = Mat::zeros(img_src.size(), img_src.type());

  for(int i = 1; i <= 255; i += 5) {
    Mat img_tmp;
    cout << i << endl;
    compare(img_src, i, img_tmp, CMP_EQ);
    bitwise_or(img_tmp, img_dst, img_dst);
  }

  imshow(win_src, img_src);
  imshow(win_dst, img_dst);

  imwrite(file_dst, img_dst);
  waitKey(0);

  return 0;
}

File: OpenCVによる画像処理入門/第3版/第2刷/sampleprogram/11-3-7.py
import cv2
import math
import numpy as np

file_src = 'src_depth_wall.png'
file_dst = 'dst.png'

img_src = cv2.imread(file_src, cv2.IMREAD_GRAYSCALE)

cv2.namedWindow('src')
cv2.namedWindow('dst')

rows, cols = img_src.shape
img_dst = np.zeros([rows, cols]).astype(np.uint8)

for i in range(1, 256, 5):
  print(i)
  img_tmp = cv2.compare(img_src, i, cv2.CMP_EQ)
  img_dst = cv2.bitwise_or(img_tmp, img_dst)

cv2.imshow('src', img_src)
cv2.imshow('dst', img_dst)

cv2.imwrite(file_dst, img_dst)
cv2.waitKey(0)
cv2.destroyAllWindows()
